src/host/readDataCooked.cpp

```diff
@@ -797,453 +807,484 @@ void COOKED_READ_DATA::_transitionState(State state) noexcept
+void COOKED_READ_DATA::_replace(const std::wstring_view& str)
+{
+    _buffer.assign(str);
+    _bufferCursor = _buffer.size();
+    _bufferDirtyBeg = 0;
+    _dirty = true;
+}
 
-    do
-    {
-        std::ignore = _writeChars({ whitespace.data(), nextWriteSize });
-        remaining -= nextWriteSize;
-        nextWriteSize = std::min(remaining, whitespace.size());
-    } while (remaining != 0);
+void COOKED_READ_DATA::_setCursorPosition(size_t position) noexcept
+{
+    _bufferCursor = std::min(position, _buffer.size());
+    _dirty = true;
 }
 
-// A helper to calculate the number of cells `text` would take up if it were written.
-// `cursorOffset` allows the caller to specify a "logical" cursor position relative to the actual cursor position.
-// This allows the function to track in which column it currently is, which is needed to implement tabs for instance.
-ptrdiff_t COOKED_READ_DATA::_measureChars(const std::wstring_view& text, ptrdiff_t cursorOffset) const
+std::wstring_view COOKED_READ_DATA::_slice(size_t from, size_t to) const noexcept
 {
-    if (text.empty())
-    {
-        return 0;
-    }
-    return _writeCharsImpl(text, true, cursorOffset);
+    to = std::min(to, _buffer.size());
+    from = std::min(from, to);
+    return std::wstring_view{ _buffer.data() + from, to - from };
 }
 
-// A helper to write text and calculate the number of cells we've written.
-// _unwindCursorPosition then allows us to go that many cells back. Tracking cells instead of explicit
-// buffer positions allows us to pay no further mind to whether the buffer scrolled up or not.
-ptrdiff_t COOKED_READ_DATA::_writeChars(const std::wstring_view& text) const
+// Draws the contents of _buffer onto the screen.
+//
+// By using the _dirty flag we avoid redrawing the buffer unless needed.
+// This turns the amortized time complexity of _readCharInputLoop() from O(n^2) (n(n+1)/2 redraws) into O(n).
+// Without this, pasting text would otherwise quickly turn into "accidentally quadratic" meme material.
+//
+// NOTE: Don't call _flushBuffer() after appending newlines to the buffer! See _handlePostCharInputLoop for more information.
+void COOKED_READ_DATA::_redisplay()
 {
-    if (text.empty())
+    if (!_dirty || WI_IsFlagClear(_pInputBuffer->InputMode, ENABLE_ECHO_INPUT))
     {
-        return 0;
+        return;
     }
-    return _writeCharsImpl(text, false, 0);
-}
 
-ptrdiff_t COOKED_READ_DATA::_writeCharsImpl(const std::wstring_view& text, const bool measureOnly, const ptrdiff_t cursorOffset) const
-{
-    const auto width = _screenInfo.GetTextBuffer().GetSize().Width();
-    auto it = text.begin();
-    const auto end = text.end();
-    ptrdiff_t distance = 0;
+    const auto size = _screenInfo.GetVtPageArea().Dimensions();
+    auto originInViewportFinal = _originInViewport;
+    til::point cursorPositionFinal;
+    til::point pagerPromptEnd;
+    std::vector<Line> lines;
 
+    // FYI: This loop does not loop. It exists because goto is considered evil

```

Comment:
well, it might loop once. which we can't know at the top of this, for the same "wide chars wrap weird" reason

Reply:
> well, it might loop once.

To me it felt like writing a duff's device. _Sure_, it does wrap around once, but like...
It literally has a `break;` at the tail end of the for-loop. 😅 For all intents and purposes, I basically wrote a `goto` but I'm not calling it as such. I just didn't want to write another `pragma suppress`.